File: Haskell/lisp.hs
module Main(toString, main, read_from, atom) where

import System.IO (hFlush, stdout)
import Debug.Trace
import Control.Exception
import Prelude hiding (catch)

data Expr = LispInt Int |
            LispSymbol String |
            LispProc ([Expr]->Expr) |
            LispLambda [Expr] |
            LispList [Expr]

instance Show Expr where
  show (LispInt x) = "Int:" ++ (show x)
  show (LispSymbol x) = "Str:" ++ x
  show (LispList x) = "List:[" ++ unwords (map show x) ++ "]"
  show (LispProc x) = "Proc:[]"
  show (LispLambda x) = "Lambda:[" ++ unwords (map show x) ++ "]"

instance Eq Expr where
  (LispInt a) == (LispInt b) = a == b
  (LispSymbol a) == (LispSymbol b) = a == b
  (LispList a) == (LispList b) = a == b
  (LispLambda a) == (LispLambda b) = a == b
  (LispProc a) == (LispProc b) = False

instance Show Environment where
  show (Environment lst parent) = "Env:" ++ (show lst) ++ ":" ++ (show parent)

data Environment = Environment [(String, Expr)] (Maybe Environment)

errorFewArgs = error "too few arguments"
errorManyArgs = error "too many arguments"


bool :: Bool -> Expr
bool True = lispTrue
bool _ = lispFalse

procAppend :: [Expr] -> Expr
procAppend [] = LispList []
procAppend (x:xs) = LispList ((toList x) ++ (toList (procAppend xs)))

procCar :: [Expr] -> Expr
procCar [] = errorFewArgs
procCar (x:xs) = head $ toList x

procCdr :: [Expr] -> Expr
procCdr [] = errorFewArgs
procCdr (x:xs) = LispList $ tail $ toList x

procCons :: [Expr] -> Expr
procCons [] = errorFewArgs
procCons (a:(LispList es):[]) = LispList (a:es)
procCons (a:b:[]) = LispList [a,b]
procCons _ = errorManyArgs

procLength :: [Expr] -> Expr
procLength ((LispList es):[]) = LispInt (length es)
procLength _ = undefined

procList :: [Expr] -> Expr
procList [] = LispList []
procList (x:xs) = LispList (x : (toList (procList xs)))

procIsList :: [Expr] -> Expr
procIsList (LispList x:xs) = lispTrue
procIsList _ = lispFalse

procIsSymbol :: [Expr] -> Expr
procIsSymbol (LispSymbol x:xs) = lispTrue
procIsSymbol _ = lispFalse

procIsNil :: [Expr] -> Expr
procIsNil (lispNil:xs) = lispTrue
procIsNil _ = lispFalse

procNot :: [Expr] -> Expr
procNot (lispTrue:xs) = lispFalse
--procNot (lispFalse:xs) = lispTrue
procNot _ = lispTrue


procPlus :: [Expr] -> Expr
procPlus exps = LispInt $ sum $ map toInt exps

procMinus :: [Expr] -> Expr
procMinus [] = errorFewArgs
procMinus (x:[]) = LispInt (- (toInt x))
procMinus (x:xs) = LispInt ((toInt x) - (sum $ map toInt xs))

procMultiply :: [Expr] -> Expr
procMultiply [] = LispInt 1
procMultiply (LispInt 0:xs) = LispInt 0
procMultiply (LispInt x:xs) = LispInt (x * (toInt $ procMultiply xs))


procDivide :: [Expr] -> Expr
procDivide [] = LispInt 1
procDivide (LispInt 0:xs) = LispInt 0
procDivide (LispInt x:xs) = LispInt (x `div` (toInt $ procMultiply xs))

procGreater :: [Expr] -> Expr
procGreater (LispInt x:xs) = bool $ all (\i -> x > i) (map toInt xs)

procLess :: [Expr] -> Expr
procLess (LispInt x:xs) = bool $ all (\i -> x < i) (map toInt xs)

procGreaterEqual :: [Expr] -> Expr
procGreaterEqual (LispInt x:xs) = bool $ all (\i -> x >= i) (map toInt xs)

procLessEqual :: [Expr] -> Expr
procLessEqual (LispInt x:xs) = bool $ all (\i -> x <= i) (map toInt xs)

procEqual :: [Expr] -> Expr
procEqual (LispInt x:xs) = bool $ all (\i -> x == i) (map toInt xs)


globalEnvironment = Environment [
  ("append", LispProc procAppend),
  ("car", LispProc procCar),
  ("cdr", LispProc procCdr),
  ("cons", LispProc procCons),
  ("length", LispProc procLength),
  ("list", LispProc procList),

  ("list?", LispProc procIsList),
  ("symbol?", LispProc procIsSymbol),
  ("nil?", LispProc procIsNil),
  ("not", LispProc procNot),

  ("+", LispProc procPlus),
  ("-", LispProc procMinus),
  ("*", LispProc procMultiply),
  ("/", LispProc procDivide),

  (">", LispProc procGreater),
  ("<", LispProc procLess),
  (">=", LispProc procGreaterEqual),
  ("<=", LispProc procLessEqual),
  ("=", LispProc procEqual),

  ("t",lispTrue),
  ("f",lispFalse),
  ("nil", lispNil)] Nothing


dict :: Environment -> [(String, Expr)]
dict (Environment lst parent) = lst

setDict :: Environment -> String -> Expr -> Environment
setDict env s e = case findEnv (Just env) s of
  (Just (Environment lst parent)) -> Environment ((s,e):lst) parent
  Nothing -> let Environment l p = env in Environment ((s,e):l) p

defineDict (Environment lst parent) s e = Environment ((s,e):lst) parent


searchExpr :: [(String,Expr)] -> String -> Maybe Expr
searchExpr ((s,expr):lst) key = if s == key then Just expr else (searchExpr lst key)
searchExpr [] key = Nothing

findEnv :: Maybe Environment -> String -> Maybe Environment
findEnv Nothing _ = Nothing
findEnv (Just (Environment lst parent)) s = case (searchExpr lst s) of
                                            Just expr -> Just (Environment lst parent)
                                            Nothing -> findEnv parent s


headExpr :: Expr -> Expr
headExpr (LispList (x:xs)) = x
headExpr (LispLambda (x:xs)) = x
headExpr _ = atom "headExpr error!!"

tailExpr :: Expr -> [Expr]
tailExpr (LispList (x:xs)) = xs
tailExpr (LispLambda (x:xs)) = xs
tailExpr _ = [atom "headExpr error!!"]

toList :: Expr -> [Expr]
toList (LispList lst) = lst
toList _ = undefined

toSym :: Expr -> String
toSym (LispSymbol s) = s
toSym _ = undefined



eval :: Environment -> Expr -> (Environment, Expr)
eval env (LispSymbol s) = case findEnv (Just env) s of
                              Just e -> case searchExpr (dict e) s of
                                        Just expr -> (env, expr)
                                        -- ここには来ないはず
                                        Nothing -> (env, atom "NO!")
                              Nothing -> (env, atom "no!?")
eval env (LispInt i) = (env, LispInt i)
eval env (LispList []) = (env, lispNil)
eval env (LispList (LispSymbol "quote":xs)) = (env, head xs)
eval env (LispList (LispSymbol "if":xs)) = case xs of
    (cond:a:b:_) -> if lispTrue == (snd (eval env cond)) then eval env a else eval env b
    _ -> (env, lispNil)
eval env (LispList (LispSymbol "set!":xs)) = case xs of
    ((LispSymbol key):cell:_) -> let c = snd (eval env cell) in ((setDict env key c), c)
    _ -> (env, lispNil)
eval env (LispList (LispSymbol "define":xs)) = case xs of
    ((LispSymbol key):cell:_) -> let c = snd (eval env cell) in ((defineDict env key c), c)
    _ -> (env, lispNil)
eval env (LispList (LispSymbol "lambda":xs)) = (env, LispLambda (LispSymbol "lambda":xs))
eval env lst = let  proc = snd $ eval env (headExpr lst) ;
                    exps = map (snd . (eval env)) (tailExpr lst)
              in case proc of
                (LispLambda l) -> let syms = map toSym (toList (head (tailExpr proc)))
                                 in eval (Environment (zip syms exps) (Just env)) (head (tail (tailExpr proc)))
                (LispProc p) -> (env, p exps)
                _ -> (env, atom ("undefined eval:" ++ (show proc)))

atom :: String -> Expr
atom s = case reads s of
          [(x, "")] -> LispInt x
          _ -> LispSymbol s

lispNil = LispSymbol "nil"
lispTrue = LispSymbol "t"
lispFalse = LispSymbol "f"

read_list :: [Expr] -> [String] -> ([Expr], [String])
read_list exprs (")":xs) = (reverse exprs, xs)
read_list exprs (x:xs) = let (expr, ss) = read_from (x:xs)
                          in read_list (expr:exprs) ss

read_from :: [String] -> (Expr, [String])
read_from [] = error "unexpected EOF while readling"
read_from ("(":xs) = let (list, ss) = read_list [] xs
                      in (LispList list, ss)
read_from (")":xs) = error "unexpedted )"
read_from (x:xs) = (atom x, xs)


tokenize :: String -> [String]
tokenize [] = []
tokenize s = words $ addSpace s
  where
    addSpace [] = []
    addSpace (x:xs) = if x == '(' || x == ')'
                      then ' ':x:' ':(addSpace xs)
                      else x:(addSpace xs)

parse :: String -> Expr
parse str = fst (read_from (tokenize str))

toString :: Expr -> String
toString = show

toInt :: Expr -> Int
toInt exp = case exp of
             LispInt i -> i
             _ -> error "not int"

repl env = do
  putStr "MyLisp> "
  hFlush stdout
  x <- getLine
  putStr "-> "
  let (e,expr) = eval env $ parse x
  putStrLn $ toString expr
  repl e

main = repl globalEnvironment

File: OCaml/lisp.ml
#load "str.cma"

let join = String.concat ","
let zip a b = List.map2 (fun x y -> x,y) a b
let sum lst = List.fold_left (fun s i -> s + i) 0 lst
let rec all f arr =
  match arr with
  | [] -> true
  | x :: xs -> if (f x) then all f xs else false
 

type sexp =
      | LispInt of int
      | LispSymbol of string
      | LispProc of (sexp list -> sexp)
      | LispLambda of (sexp list)
      | LispList of (sexp list)

type environment = Environment of (((string * sexp) list) * (environment option))

let lisp_nil = LispSymbol "nil"
let lisp_true = LispSymbol "t"
let lisp_false = LispSymbol "f"
let lisp_error = LispSymbol "error"

let lisp_bool cond = if cond then lisp_true else lisp_false


let atom s =
  try LispInt (int_of_string s)
  with _ -> LispSymbol s

let rec to_sym sexp =
  match sexp with
  | LispSymbol s -> s
  | a -> to_sym lisp_error

let to_list sexp =
  match sexp with
  | LispList lst -> lst
  | a -> [lisp_nil]

let to_int sexp =
  match sexp with
  | LispInt i -> i
  | a -> 0

let rec to_string sexp =
  match sexp with
  | LispInt i -> "Int(" ^ (string_of_int i) ^ ")"
  | LispSymbol s -> "Sym(\"" ^ s ^ "\")"
  | LispProc lst -> "Proc[]"
  | LispList lst -> "List[" ^ (join (List.map to_string lst)) ^ "]"
  | LispLambda lst -> "Lambda[]"

let error_few_args = atom "too few arguments"
let error_many_args = atom "too many arguments"
let error_undefined = atom "undefined"

let rec proc_append lst =
  match lst with
  | [] -> LispList []
  | x::xs -> LispList (List.append (to_list x) (to_list (proc_append xs)))

let proc_car lst =
  match lst with
  | [] -> error_few_args
  | x :: xs -> List.hd (to_list x)

let proc_cdr lst =
  match lst with
  | [] -> error_few_args
  | x::xs -> LispList (List.tl (to_list x))

let proc_cons lst =
  match lst with
  | [] -> error_few_args
  | a::(LispList es)::[] -> LispList (a::es)
  | a::b::[] -> LispList [a;b]
  | _ -> error_many_args

let proc_length lst =
  match lst with
  | (LispList es)::[] -> LispInt (List.length es)
  | _ -> error_undefined

let rec proc_list lst =
  match lst with
  | [] -> LispList []
  | x::xs -> LispList (x :: (to_list (proc_list xs)))

let proc_is_list lst =
  match lst with
  | (LispList x)::xs -> lisp_true
  | _ -> lisp_false

let proc_is_symbol lst =
  match lst with
  | (LispSymbol x)::xs -> lisp_true
  | _ -> lisp_false

let proc_is_nil lst = lisp_bool ((List.hd lst) == lisp_nil)
let proc_not lst = lisp_bool ((List.hd lst) == lisp_true)

let proc_plus lst = LispInt (sum (List.map to_int lst))

let proc_minus lst =
  match lst with
  | (LispInt x)::[] -> LispInt (- x)
  | (LispInt x)::xs -> LispInt (x - (sum (List.map to_int xs)))
  | _ -> error_few_args

let rec proc_multiply lst =
  match lst with
  | [] -> LispInt 1
  | (LispInt 0) :: xs -> LispInt 0
  | (LispInt x) :: xs -> LispInt (x * (to_int (proc_multiply xs)))
  | _ -> error_undefined

let proc_divide lst =
  match lst with
  | [] -> LispInt 1
  | (LispInt 0)::xs -> LispInt 0
  | (LispInt x)::xs -> LispInt (x / (to_int (proc_multiply xs)))
  | _ -> error_undefined


let compare_all f lst =
   match lst with
  | (LispInt x)::xs -> lisp_bool (all (fun i -> (f x i)) (List.map to_int xs))
  | _ -> error_undefined

let proc_greater = compare_all (fun x i -> x > i)
let proc_less = compare_all (fun x i -> x < i)
let proc_greater_equal = compare_all (fun x i -> x >= i)
let proc_less_equal = compare_all (fun x i -> x <= i)
let proc_equal = compare_all (fun x i -> x == i)

let global_environment =
    Environment ([
      ("append", LispProc proc_append);
      ("car", LispProc proc_car);
      ("cdr", LispProc proc_cdr);
      ("cons", LispProc proc_cons);
      ("length", LispProc proc_length);
      ("list", LispProc proc_list);

      ("list?", LispProc proc_is_list);
      ("symbol?", LispProc proc_is_symbol);
      ("nil?", LispProc proc_is_nil);
      ("not", LispProc proc_not);
      
      ("+", LispProc proc_plus);
      ("-", LispProc proc_minus);
      ("*", LispProc proc_multiply);
      ("/", LispProc proc_divide);

      (">", LispProc proc_greater);
      ("<", LispProc proc_less);
      (">=", LispProc proc_greater_equal);
      ("<=", LispProc proc_less_equal);
      ("=", LispProc proc_equal);

      ("t",lisp_true);
      ("f",lisp_false);
      ("nil",lisp_nil)
    ], None)


let rec search_dict lst sym =
  match lst with
  | [] -> None
  | (s,e)::rest ->
      if s = sym then
        Some e
      else
        search_dict rest sym

let rec search_env env sym =
  match env with
  | None -> (None, lisp_nil)
  | Some Environment (lst, parent) ->
      match (search_dict lst sym) with
      | None -> search_env parent sym
      | Some exp -> (Some (Environment (lst, parent)), exp)

let define_dict env sym sexp =
  let Environment (lst, parent) = env
  in Environment (((sym,sexp)::lst), parent)

let set_dict env sym sexp =
  match search_env (Some env) sym with
  | (None, a) -> define_dict env sym sexp
  | (Some e, a) -> define_dict e sym sexp



let rec eval env sexp =
  let
    eval_proc env sexp =
      match List.map (fun x -> snd (eval env x)) (to_list sexp) with
      | [] -> (env, lisp_error)
      | (proc::exps) -> 
          match proc with
          | LispLambda (l::args::body::a) ->
              let syms = List.map to_sym (to_list args)
              in eval (Environment ((zip syms exps), Some env)) body
          | LispProc p -> (env, p exps)
          | a -> (env, atom "undefined eval")
  in
  let
    eval_list_quote lst = List.hd lst
  in
  let
    eval_list_if lst =
      match lst with
      | (cond::a::b::_) -> if lisp_true == (snd (eval env cond)) then eval env a else eval env a
      | _ -> (env, lisp_nil)
  in
  let
    eval_list_set lst =
      match lst with
      | ((LispSymbol key)::cell::_) -> let c = snd (eval env cell) in ((set_dict env key c), c)
      | _ -> (env, lisp_nil)
  in
  let
    eval_list_define lst =
      match lst with
      | ((LispSymbol key)::cell::_) -> let c = snd (eval env cell) in ((define_dict env key c), c)
      | _ -> (env, lisp_nil)
  in
  let
    eval_list env sexp =
      let lst = to_list sexp in
      let rest = List.tl lst in
      match to_sym (List.hd lst) with
      | "quote" -> (env, eval_list_quote rest)
      | "if" -> eval_list_if rest
      | "set!" -> eval_list_set rest
      | "define" -> eval_list_define rest
      | "lambda" -> (env, LispLambda (to_list sexp))
      | _ -> eval_proc env sexp
  in
  match sexp with
  | LispSymbol sym -> (env, snd (search_env (Some env) sym))
  | LispInt _ -> (env, sexp)
  | LispList [] -> (env, lisp_nil)
  | LispList _ -> eval_list env sexp
  | _ -> eval_proc env sexp

let rec read_from str_list =
  let
    rec read_list exprs lst =
      match lst with
      | [] -> [lisp_nil], []
      | ")"::xs -> (List.rev exprs), xs
      | x::xs -> let (expr, ss) = read_from (x::xs) in read_list (expr::exprs) ss
  in
    match str_list with
    | [] -> (lisp_nil, [])
    | ("("::xs) -> let (lst, ss) = read_list [] xs in (LispList lst, ss)
    | (")"::xs) -> (lisp_nil, [])
    | (x::xs) -> (atom x, xs)


let tokenize s =
  let add_space = Str.global_replace (Str.regexp "[()]") " \\0 "
  in Str.split (Str.regexp " +") (String.trim (add_space s))

let parse s = fst (read_from (tokenize s))

let rec repl env = (

  print_string "MyLisp> ";
  flush stdout;
  let
    s = input_line stdin
  in
  let
    (e,sexp) = eval env (parse s)
  in
    print_string "-> ";
    print_string (to_string sexp);
    print_string "\n";

    flush stdout;
    repl e;
)

let _ = repl global_environment
